src/IREquality.cpp

```diff
@@ -10,725 +10,558 @@ using std::vector;
+        cmp(a.min, b.min);
+        cmp(a.extent, b.extent);
+    }
+
+    HALIDE_ALWAYS_INLINE
+    void cmp(const ModulusRemainder &a, const ModulusRemainder &b) {
+        cmp(a.modulus, b.modulus);
+        cmp(a.remainder, b.remainder);
+    }
+
+    void cmp(const halide_handle_cplusplus_type *ha,
+             const halide_handle_cplusplus_type *hb) {
+        if (ha == hb) {
+            return;
+        } else if (!ha) {
+            result = Order::LessThan;
+        } else if (!hb) {
+            result = Order::GreaterThan;
+        } else {
+            // They're both non-void handle types with distinct type info
+            // structs. We now need to distinguish between different C++
+            // pointer types (e.g. char * vs const float *). If would be nice
+            // if the structs were unique per C++ type. Then comparing the
+            // pointers above would be sufficient.  Unfortunately, different
+            // shared libraries in the same process each create a distinct
+            // struct for the same type. We therefore have to do a deep
+            // comparison of the type info fields.
+            cmp(ha->reference_type, hb->reference_type);
+            cmp(ha->inner_name.name, hb->inner_name.name);
+            cmp(ha->inner_name.cpp_type_type, hb->inner_name.cpp_type_type);
+            cmp(ha->namespaces, hb->namespaces);
+            cmp(ha->enclosing_types, hb->enclosing_types);
+            cmp(ha->cpp_type_modifiers, hb->cpp_type_modifiers);
+        }
     }
 
-    for (size_t i = 0; i < ha->enclosing_types.size(); i++) {
-        compare_scalar(ha->enclosing_types[i].cpp_type_type,
-                       hb->enclosing_types[i].cpp_type_type);
-        compare_names(ha->enclosing_types[i].name,
-                      hb->enclosing_types[i].name);
+    HALIDE_ALWAYS_INLINE
+    void cmp(const Type &a, const Type &b) {
+        uint32_t ta = ((halide_type_t)a).as_u32();
+        uint32_t tb = ((halide_type_t)b).as_u32();
+        if (ta < tb) {
+            result = Order::LessThan;
+        } else if (ta > tb) {
+            result = Order::GreaterThan;
+        } else {
+            if (a.handle_type || b.handle_type) {
+                cmp(a.handle_type, b.handle_type);
+            }
+        }
     }
 
-    if (result != Equal) {
-        return result;
+    void cmp(const PrefetchDirective &a, const PrefetchDirective &b) {
+        cmp(a.name, b.name);
+        cmp(a.at, b.at);
+        cmp(a.from, b.from);
+        cmp(a.offset, b.offset);
+        cmp(a.strategy, b.strategy);
     }
 
-    for (size_t i = 0; i < ha->cpp_type_modifiers.size(); i++) {
-        compare_scalar(ha->cpp_type_modifiers[i],
-                       hb->cpp_type_modifiers[i]);
+    HALIDE_ALWAYS_INLINE
+    void cmp(double a, double b) {
+        // Floating point scalars need special handling, due to NaNs.
+        if (std::isnan(a) && std::isnan(b)) {
+        } else if (std::isnan(a)) {
```

Comment:
Did you mean to leave the both-nan case empty? If so, please add a comment like `/*deliberately empty, leave comparison state unchanged */`